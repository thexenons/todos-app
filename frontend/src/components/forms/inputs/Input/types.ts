import type { InputHTMLAttributes } from "react";
import { FieldValues, RegisterOptions } from "react-hook-form";

export interface InputBaseProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
}

export interface InputProps extends InputBaseProps {
  name: string;
  rules?: Omit<
    RegisterOptions<FieldValues, string>,
    "valueAsNumber" | "valueAsDate" | "setValueAs" | "disabled"
  >;
}
