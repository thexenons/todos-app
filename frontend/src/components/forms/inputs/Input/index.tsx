import { forwardRef, useId } from "react";
import { useController, useFormContext } from "react-hook-form";
import * as S from "./styled";
import type { InputBaseProps, InputProps } from "./types";

export const InputBase = forwardRef<HTMLInputElement, InputBaseProps>(
	({ label = "", ...rest }, ref) => {
		const id = useId();

		return (
			<S.InputWrapper>
				<S.InputLabel htmlFor={rest.id || id}>{label}</S.InputLabel>
				<S.InputElement {...rest} id={rest.id || id} ref={ref} />
			</S.InputWrapper>
		);
	}
);
InputBase.displayName = "InputBase";

const Input = forwardRef<HTMLInputElement, InputProps>(
	({ rules, name, ...rest }, ref) => {
		const { control } = useFormContext();
		const { field } = useController({
			name,
			control,
			rules,
			defaultValue: rest.value || "",
		});
		const handleRef = (inputRef: HTMLInputElement) => {
			field.ref(inputRef);
			if (typeof ref === "function") {
				ref(inputRef);
			}
		};

		return <InputBase {...rest} {...field} ref={handleRef} />;
	}
);
Input.displayName = "Input";

export default Input;
