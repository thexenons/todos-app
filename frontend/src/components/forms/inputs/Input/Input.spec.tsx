import { render } from "@testing-library/react";

import InputTestComponent from "../../../../tests/InputTestComponent";
import Input from ".";

describe("Input", () => {
	it("Render", () => {
		const { container } = render(
			<InputTestComponent>
				<Input name="test" />
			</InputTestComponent>
		);
		expect(container).toMatchSnapshot();
	});

	it("should set ref", () => {
		let inputElement: HTMLInputElement | undefined;
		const handleRef = (ref: HTMLInputElement) => {
			inputElement = ref;
		};
		render(
			<InputTestComponent>
				<Input name="test" ref={handleRef} />
			</InputTestComponent>
		);
		expect(inputElement).not.toBeNull();
	});
});
