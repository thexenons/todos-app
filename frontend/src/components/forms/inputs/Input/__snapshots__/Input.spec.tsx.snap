// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input > Render 1`] = `
<div>
  <form>
    <div
      class="styled__InputWrapper-sc-f06926e2-2 bMkOPf"
    >
      <input
        class="styled__InputElement-sc-f06926e2-1 ekMEKq"
        data-has-placeholder="false"
        id=":r0:"
        name="test"
        placeholder=" "
        type="text"
        value=""
      />
      <label
        class="styled__InputLabel-sc-f06926e2-0 eomqRx"
        for=":r0:"
      />
    </div>
  </form>
</div>
`;
